packages/flutter_tools/lib/src/migrate/migrate_utils.dart

```diff
@@ -0,0 +1,325 @@
+    }
+    return outputDirectory;
+  }
+
+  /// Runs the git 3-way merge on three files and returns the results as a MergeResult.
+  ///
+  /// Passing the same path for base and current will perform a two-way fast forward merge.
+  static Future<MergeResult> gitMergeFile({
+    required String base,
+    required String current,
+    required String target,
+    required String localPath,
+    required Logger logger
+  }) async {
+    final List<String> cmdArgs = <String>['merge-file', '-p', current, base, target];
+    final ProcessResult result = await Process.run('git', cmdArgs);
+    checkForErrors(result, logger, allowedExitCodes: <int>[-1], commandDescription: 'git ${cmdArgs.join(' ')}');
+    return MergeResult(result, localPath);
+  }
+
+  /// Calls `git init` on the workingDirectory.
+  static Future<void> gitInit(String workingDirectory, Logger logger) async {
+    final List<String> cmdArgs = <String>['init'];
+    final ProcessResult result = await Process.run('git', cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, logger, allowedExitCodes: <int>[0], commandDescription: 'git ${cmdArgs.join(' ')}');
+  }
+
+  /// Returns true if the workingDirectory git repo has any uncommited changes.
+  static Future<bool> hasUncommitedChanges(String workingDirectory, Logger logger) async {
+    final List<String> cmdArgs = <String>['diff', '--quiet', 'HEAD', '--', '.', "':(exclude)$kDefaultMigrateWorkingDirectoryName'"];
+    final ProcessResult result = await Process.run('git', cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, logger, allowedExitCodes: <int>[-1], commandDescription: 'git ${cmdArgs.join(' ')}');
+    if (result.exitCode == 0) {
+      return false;
+    }
+    return true;
+  }
+
+  /// Returns true if the workingDirectory is a git repo.
+  static Future<bool> isGitRepo(String workingDirectory, Logger logger) async {
+    final List<String> cmdArgs = <String>['rev-parse', '--is-inside-work-tree'];
+    final ProcessResult result = await Process.run('git', cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, logger, allowedExitCodes: <int>[-1], commandDescription: 'git ${cmdArgs.join(' ')}');
+    if (result.exitCode == 0) {
+      return true;
+    }
+    return false;
+  }
+
+  /// Returns true if the file at `filePath` is covered by the `.gitignore`
+  static Future<bool> isGitIgnored(String filePath, String workingDirectory, Logger logger) async {
+    final List<String> cmdArgs = <String>['check-ignore', filePath];
+    final ProcessResult result = await Process.run('git', cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, logger, allowedExitCodes: <int>[0, 1, 128], commandDescription: 'git ${cmdArgs.join(' ')}');
+    return result.exitCode == 0;
+  }
+
+  /// Runs `flutter pub upgrate --major-revisions`.
+  static Future<void> flutterPubUpgrade(String workingDirectory, Logger logger) async {
```

Comment:
Why do we need this?

Reply:
This is part of an optional step in `flutter migrate apply` that recommends developers to update dependency locks if detected. Without this, it is likely their apps will not build. Here, we can handle it for them to make this more of a complete experience. See https://github.com/flutter/flutter/pull/100284/files#diff-b220b86cd875c0245243136047cb1cb1ea5cc790fe196eafa2fe29239a2ea4f7R172

Reply:
Why wouldn't their app build? Is it because the Flutter SDK changed and changed some shared deps?